### PR TITLE
Get sync in order before adding TChannel transport

### PR DESCRIPTION
**Pull request by breerly**
This PR brings sync back from the dead before we add the TChannel transport by:

- Removing sync server support.
- Using threadloop so we automatically get support for all outbounds and encodings.
- Adding sync to cross language tests.

TODO:
- [x] remove sync server support
- [x] use threadloop instead of reimplementing transports
- [x] add sync to crossdock test suite
- [x] fix examples dir
- [x] remove debug logging
- [x] remove coroutines (this is why lint is failing ATM)
- [x] raw test
- [x] json test
- [x] thrift test
- [x] fix callback exception handling (use fail_to from tchannel-python).
- [x] use custom exception for sync register
- [x] rebase

@yarpc/yarpc @yarpc/python

### CODE REVIEW

File: crossdock/main.py [hunk 1]
```diff
@@ -2,16 +2,27 @@
...
+    if not SYNC:
+        server.start()
     client.start()
 
 if __name__ == '__main__':
     main()
-    tornado.ioloop.IOLoop.current().start()
+    if not SYNC:
+        import tornado
+        tornado.ioloop.IOLoop.current().start()
```
Comment (abhinav) [inline crossdock/main.py]: Given the amount of branching and the size of this file, I would just write two `main` functions. It'll be a lot easier.

File: crossdock/sync/client/client.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    'echo': echo,
+}
+
+PASSED = 'passed'
+SKIPPED = 'skipped'
+FAILED = 'failed'
+
+
+def start():
+    """Start yarpc-python-sync test harness."""
```
Comment (abhinav) [inline crossdock/sync/client/client.py]: Usually app instantiation and route registration for Flask apps happens at the top level of the module. 

```python
app = Flask(__name__)

@app.route(...)
...

def start():
    app.run(...)
```

File: crossdock/sync/client/client.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def ready():
+        """Crossdock makes a HEAD request to see when client is ready."""
+        pass
+
+    @app.route('/', methods=['GET'])
+    def handle():
+        """Crossdock sends GET requests with query params to initiate test."""
+
+        behavior = request.args.get('behavior')
+        server = request.args.get('server')
```
Comment (abhinav) [inline crossdock/sync/client/client.py]: `get` will return None if the parameters are absent. Validate their presence or use `request.args['behavior']` to make this fail if the parameters are absent.

File: crossdock/sync/client/echo.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
+
+from crossdock.async.client.echo import rand_str
```
Comment (abhinav) [inline crossdock/sync/client/echo.py]: Don't import from the async module here. Either put this in a shared place or just re-implement it.

File: crossdock/sync/client/echo.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    expected = {
+        'token': rand_str(7),
+    }
+
+    request = Request(
+        service='yarpc-test',
+        procedure='echo',
+        body=expected,
+        hostport=server,
```
Comment (abhinav) [inline crossdock/sync/client/echo.py]: Do you have a TODO/issue somewhere to move the hostport to the HTTPOutbound?

Comment (breerly) [inline crossdock/sync/client/echo.py]: yup

Comment (breerly) [inline crossdock/sync/client/echo.py]: https://github.com/yarpc/yarpc-python/issues/25

File: yarpc/rpc.py [hunk 6]
```diff
@@ -30,6 +34,8 @@ def register(self, procedure, handler):
         self._dispatcher.register(procedure, handler)
 
     def start(self):
+        log.info('starting')
+
```
Comment (abhinav) [inline yarpc/rpc.py]: Did you mean to leave this here?

File: crossdock/sync/client/echo.py (hunk 5 above)
Comment (abhinav) [inline crossdock/sync/client/echo.py]: :+1:

File: yarpc/rpc.py (hunk 6 above)
Comment (breerly) [inline yarpc/rpc.py]: nope - thx

File: crossdock/sync/client/echo.py (hunk 4 above)
Comment (breerly) [inline crossdock/sync/client/echo.py]: ill move it up

File: crossdock/main.py (hunk 1 above)
Comment (breerly) [inline crossdock/main.py]: yeah - i kind of hacked this in. let me spruce it up.

File: crossdock/sync/client/client.py (hunk 2 above)
Comment (breerly) [inline crossdock/sync/client/client.py]: yeah sure

File: yarpc/rpc.py [hunk 7]
```diff
@@ -43,6 +49,6 @@ def call(self, request):
         request.caller = self._service
 
         outbound = self._outbounds.get(request.service)[0]
```
Comment (junchaowu) [inline yarpc/rpc.py]: _outbounds.get() return a list?

Comment (breerly) [inline yarpc/rpc.py]: Yeah - there is a ticket to switching that to a single outbound.

Comment (breerly) [inline yarpc/rpc.py]: https://github.com/yarpc/yarpc-python/issues/10

**Comment by breerly** [comment]
@abhinav thanks for the feedback, addressed now :)

File: crossdock/rand.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
+
+
+def string(length):
+    from random import choice
+    from string import ascii_lowercase
```
Comment (abhinav) [inline crossdock/rand.py]: Why are these imports inside the function? There's no cyclic imports issue here.

**Comment by abhinav** [comment]
Minor nit. LGTM otherwise. :shipit:

File: crossdock/rand.py (hunk 8 above)
Comment (breerly) [inline crossdock/rand.py]: uuuuum :+1:

**Comment by junchaowu** [comment]
lgtm

**Pull request merged**